File: python/src/warehouse/adapters/ids.py
"""Deterministic and sequence-based order id generators."""

from __future__ import annotations

from warehouse.domain.order import OrderId


class SequenceOrderIdGenerator:
    """Test double that issues ``ord-1``, ``ord-2``, ..."""

    def __init__(self, prefix: str = "ord") -> None:
        """Start at zero so the first id is prefix-1."""
        self._n = 0
        self._prefix = prefix

    def next(self) -> OrderId:
        """Return the next sequenced identifier."""
        self._n += 1
        return OrderId(value=f"{self._prefix}-{self._n}")


class FixedOrderIdGenerator:
    """Test double that always returns the same injected identifier."""

    def __init__(self, order_id: OrderId) -> None:
        """Hold a single identifier."""
        self._order_id = order_id

    def next(self) -> OrderId:
        """Return the configured identifier."""
        return self._order_id

File: typescript/src/adapters/ids.ts
/**
 * Deterministic and sequence-based order id generators.
 */

import type { OrderIdGenerator } from "../application/ports";
import { createOrderId, type OrderId } from "../domain/order";

/**
 * Test double that issues `ord-1`, `ord-2`, ...
 */
export class SequenceOrderIdGenerator implements OrderIdGenerator {
  private n = 0;

  private readonly prefix: string;

  /**
   * Starts at zero so the first id is prefix-1.
   *
   * @param prefix - leading text of each issued identifier.
   */
  constructor(prefix = "ord") {
    this.prefix = prefix;
  }

  /**
   * Returns the next sequenced identifier.
   */
  next(): OrderId {
    this.n += 1;
    return createOrderId(`${this.prefix}-${String(this.n)}`);
  }
}

/**
 * Test double that always returns the same injected identifier.
 */
export class FixedOrderIdGenerator implements OrderIdGenerator {
  private readonly orderId: OrderId;

  /**
   * Holds a single identifier.
   *
   * @param orderId - the identifier every `next` call returns.
   */
  constructor(orderId: OrderId) {
    this.orderId = orderId;
  }

  /**
   * Returns the configured identifier.
   */
  next(): OrderId {
    return this.orderId;
  }
}
